File: DSA/Graph/Graph.cpp
#include <iostream>
#include <vector>
using namespace std;

class Graph {
private:
    int num_vertices;
    vector<vector<int>> adj_list;

public:
    // Constructor to initialize the graph with given number of vertices
    Graph(int vertices) {
        num_vertices = vertices;
        adj_list.resize(num_vertices);
    }

    // Function to add an undirected edge between two vertices
    void addEdge(int vertex1, int vertex2) {
        // Adding vertex2 to the adjacency list of vertex1
        adj_list[vertex1].push_back(vertex2);
        // Adding vertex1 to the adjacency list of vertex2
        adj_list[vertex2].push_back(vertex1);
    }

    // Function to remove an undirected edge between two vertices
    void removeEdge(int vertex1, int vertex2) {
        // Finding the position of vertex2 in the adjacency list of vertex1
        int position = -1;
        for (int i = 0; i < adj_list[vertex1].size(); i++) {
            if (adj_list[vertex1][i] == vertex2) {
                position = i;
                break;
            }
        }
        if (position != -1) {
            // Removing vertex2 from the adjacency list of vertex1
            adj_list[vertex1].erase(adj_list[vertex1].begin() + position);
        }

        // Finding the position of vertex1 in the adjacency list of vertex2
        position = -1;
        for (int i = 0; i < adj_list[vertex2].size(); i++) {
            if (adj_list[vertex2][i] == vertex1) {
                position = i;
                break;
            }
        }
        if (position != -1) {
            // Removing vertex1 from the adjacency list of vertex2
            adj_list[vertex2].erase(adj_list[vertex2].begin() + position);
        }
    }

    // Function to display the adjacency list of each vertex in the graph
    void display() {
        for (int vertex = 0; vertex < num_vertices; vertex++) {
            cout << vertex << " -> ";
            for (int i = 0; i < adj_list[vertex].size(); i++) {
                cout << adj_list[vertex][i] << " ";
            }
            cout << endl;
        }
    }
};

int main() {
    int num_vertices, choice, vertex1, vertex2;
    cout << "Enter the number of vertices: ";
    cin >> num_vertices;
    Graph graph(num_vertices);

    do {
        cout << "\n1. Add edge\n2. Remove edge\n3. Display graph\n4. Quit\nEnter your choice: ";
        cin >> choice;
        switch (choice) {
            case 1:
                cout << "Enter the two vertices to connect: ";
                cin >> vertex1 >> vertex2;
                graph.addEdge(vertex1, vertex2);
                break;
            case 2:
                cout << "Enter the two vertices to disconnect: ";
                cin >> vertex1 >> vertex2;
                graph.removeEdge(vertex1, vertex2);
                break;
            case 3:
                graph.display();
                break;
            case 4:
                break;
            default:
                cout << "Invalid choice\n";
        }
    } while (choice != 4);

    return 0;
}

File: DSA/Graph/Graph.java
import java.util.Scanner;
import java.util.InputMismatchException;
class CustomGraph {
  int vertex;
  int edge;
  int[][] graph;
  Scanner sc = new Scanner(System.in);

  CustomGraph(int vertex, int edge) //constructor to initialize vertex and edges
  {
    this.vertex = vertex;
    this.edge = edge;
    this.graph = new int[vertex + 1][vertex + 1];
  }

  void insert(int vertex1, int vertex2, int weight) // to create adjacency matrix 
  {
    graph[vertex1][vertex2] = weight;
    graph[vertex2][vertex1] = weight;
  }

  void adjMaxtrix() // to initialize adjacency matrix
  {
    for (int index = 0; index < edge; index++) {
      System.out.println("enter starting vertex ");
      int value1 = sc.nextInt();
      System.out.println("enter ending vertex ");
      int value2 = sc.nextInt();
      insert(value1, value2, 1);
      System.out.println("==================");
    }
  }

  void printMatrix() //to print adjacency matrix
  {
    for (int row = 0; row < vertex; row++) {
      for (int column = 0; column < vertex; column++)
        System.out.print(graph[row][column] + "  ");

      System.out.println();
    }
  }

  void dfs(int start, boolean[] visited) {
    // Print the current node
    System.out.print(start + " ");
    // Set current node as visited
    visited[start] = true;
    // For every node of the graph
    for (int index = 0; index < graph[start].length; index++) {
      // If some node is adjacent to the current node
      // and it has not already been visited
      if (graph[start][index] == 1 && (!visited[index])) {
        dfs(index, visited);
      }
    }
  }

}
class Main {
  public static void main(String ar[]) {
    Scanner sc = new Scanner(System.in);
    try {
      System.out.println("Enter number of vertices");
      int vertex = sc.nextInt();
      System.out.println("Enter number of edges");
      int edge = sc.nextInt();
      CustomGraph g = new CustomGraph(vertex, edge);

      while (true) {
        System.out.println("===========================");
        System.out.println("*****Enter your choice*****");
        System.out.println("1. Create Adjancy matrix ");
        System.out.println("2. Print Adjancy matrix");
        System.out.println("3. Traversing DFS");
        System.out.println("4. Exit");
        System.out.println("===========================");
        try {
          int condition = sc.nextInt();
          switch (condition) {
          case 1:
            g.adjMaxtrix();
            System.out.println("Done ");
            break;

          case 2:
            g.printMatrix();
            break;

          case 3:
            boolean[] visited = new boolean[vertex];
            g.dfs(0, visited);
            System.out.println();
            break;

          case 4:
            System.out.println("Thank you");
            System.exit(0);
            break;

          default:
            System.out.println("Invalid");
            break;
          }
        } catch (Exception e) {
          System.out.println("Enter valid option");
        }
      }
    } catch (InputMismatchException e) {
      System.out.println("Enter valid option");
    }

  }
}